experimental/eval/predict.py

```diff
@@ -0,0 +1,184 @@
+                # If so, a connection can never be made.
+                retcode = self.launcher.poll()
+                if retcode is not None:
+                    raise RuntimeError(
+                        f"launcher exited unexpectedly with code {retcode}"
+                    )
+                return False
+
+        while not webserver_ready():
+            time.sleep(1.0)
+
+        print("Tabby server ready!")
+
+    def __exit__(self, _exc_type, _exc_value, _traceback):
+        self.launcher.terminate()
+
+    @method()
+    async def health(self):
+        from tabby_python_client.api.v1 import health
+
+        resp = await health.asyncio(client=self.client)
+        return resp.to_dict()
+
+    @method()
+    async def complete(self, language, crossfile_context, index, row):
+        from tabby_python_client.api.v1 import completion
+        from tabby_python_client.models import (
+            CompletionRequest,
+            DebugOptions,
+            CompletionResponse,
+            Segments,
+        )
+        from tabby_python_client.types import Response
+        from tabby_python_client import errors
+
+        if 'prediction' in row and not pd.isnull(row['prediction']):
+            return None, None, None
+
+        if crossfile_context:
+            prompt = row["crossfile_context"]["text"] + row["prompt"]
+        else:
+            prompt = row["prompt"]
+
+        groundtruth = row["groundtruth"]
+       
+        request = CompletionRequest(
+            language=language, debug_options=DebugOptions(raw_prompt=prompt)
+        )
+        # resp: CompletionResponse = await completion.asyncio(
+        #     client=self.client, json_body=request
+        # )
+        try:
+            resp: Response = await completion.asyncio_detailed(
+                client=self.client, json_body=request
+            )
+        
+            if resp.parsed != None:
+                return index, resp.parsed.choices[0].text, None
+            else:
+                return index, None, f"<{resp.status_code}>"
+        except errors.UnexpectedStatus as e:
+            return index, None, f"error: code={e.status_code} content={e.content} error={e}"
+        except Exception as e:
+            return index, None, f"error type: {type(e)}"
+
+
+
+@stub.local_entrypoint()
+async def main(language, file):
+    import json
+
+    print(MODEL_ID)
+
+    model = Model()
+    print(model.health.remote())
+
+    whole_path_file = "./data/" + MODEL_ID.split("/")[-1] + "/" + language + "/" + file
+
+    if file == 'line_completion.jsonl':
+        crossfile_context = False
+    else:
+        crossfile_context = True
+
+    objs = []
+    with open(whole_path_file) as fin:
+        for line in fin:
+            obj = json.loads(line)
+            objs.append(obj)
+
+    df = pd.DataFrame(objs)
+
+    outputs = await asyncio.gather(*[model.complete.remote.aio(language, crossfile_context, index, row) for index, row in df.iterrows()])
```

Comment:
this is fine, but you might wanna look at `return_exceptions` in https://docs.python.org/3/library/asyncio-task.html#asyncio.gather

Which could probably further simplify your implementation